### PR TITLE
Review adjustments for role restrict export

### PR DESCRIPTION
**Pull request by TheWalkingLeek**
Refs: https://github.com/hitobito/hitobito_die_mitte/issues/213

### CODE REVIEW

File: app/domain/export/tabular/people/person_row.rb [hunk 1]
```diff
@@ -31,11 +31,12 @@ def roles
...
-      if restrict_to_roles.present?
-        entry
-          .roles
-          .where(type: restrict_to_roles)
-          .where(group_id: restrict_to_group_ids)
+      if role_restrictions.present?
+        roles = entry.roles
+
+        role_restrictions.flat_map do |group_id, role_types|
+          role_types.map { |type| roles.find_by(group_id: group_id, type: type) }.compact
```
Comment (carlobeltrame) [inline app/domain/export/tabular/people/person_row.rb:38]: Müsstest du nach `flat_map` nicht noch `uniq` machen? Wenn ich zwei Filter habe die dieselbe Rolle in derselben Gruppe filtern. Bei Personenfiltern geht das nicht, aber bei Abos schon: Zum Beispiel wenn ich alle Leitenden mit dem Tag x und in einer separaten Subscription alle Leitenden und Mitglieder mit dem Tag y einschliesse.

File: app/domain/export/tabular/people/group_roles/filter.rb [hunk 2]
```diff
@@ -14,12 +14,18 @@ def initialize(group_id, list_filter_args)
...
-        restrict_to_group_ids: group_ids }
+      { role_restrictions: role_restrictions }
     end
 
     private
 
+    def role_restrictions
+      role_types = role_sti_names
+      group_ids.index_with do
+        role_types
```
Comment (carlobeltrame) [inline app/domain/export/tabular/people/group_roles/filter.rb:25]: Ich glaube das ist falsch. Die Rollentypen dürfen nicht für alle Gruppen dieselben sein.

**Review by carlobeltrame — CHANGES_REQUESTED** [review]
Hast du noch den Test für den Edge Case von https://github.com/hitobito/hitobito/pull/1886#discussion_r1012679593 ?

Bei mir hats noch nicht geklappt beim manuellen Testen, ich bekomme immer noch alle Rollen auch wenn der Haken gesetzt ist. Aber vielleicht habe ich es noch nicht richtig hingekriegt mit dem Test Setup.

File: app/domain/export/tabular/people/group_roles/mailing_list.rb [hunk 3]
```diff
@@ -13,22 +13,15 @@ def initialize(mailing_list)
...
 
     private
 
-    def role_sti_names
-      sti_names = []
-      group_subscriptions.each do |s|
-        sti_names += s.role_types
+    def role_restrictions
+      group_subscriptions.each_with_object({}) do |s, restrictions|
+        restrictions[s.subscriber_id] = s.role_types
```
Comment (carlobeltrame) [inline app/domain/export/tabular/people/group_roles/mailing_list.rb:23]: Das überschreibt die `role_types` der Gruppe, wenn dieselbe Gruppe mehr als einmal vorkommt, nicht?

**Comment by carlobeltrame** [comment]
Mir ist gerade noch aufgefallen dass die Edge Cases noch tiefer gehen:

1. Ich habe eine Gruppe vom Typ Arbeitsgruppe, nennen wir sie Gruppe A
2. Ich erstelle eine Person X mit der Rolle "Leitung" in Gruppe A und vergebe der Person X das Tag "test"
3. Ich erstelle eine Person Y mit den Rollen "Leitung" und "Mitglied" in Gruppe A
4. Ich erstelle ein Abo
5. Im Abo füge ich eine Gruppe / Rolle "Leitung" in Gruppe A hinzu, die aber nur mit Tag "test" einschliesst
6. Im selben Abo füge ich eine Gruppe / Rolle "Mitglied" hinzu
7. Ich exportiere die Abonnenten mit der neuen Checkbox aktiviert

**Erwartetes Ergebnis**: Ich sehe auf Person X die Rolle "Leitung Gruppe A" (weil Person X das Tag "test" hat) bei Person Y die Rolle "Mitglied Gruppe A" im Export (weil die Person Y kein Tag "test" hat, war ihre Leitungsfunktion nicht relevant für das Abo).
**Tatsächliches Ergebnis basierend auf der aktuellen Implementation**: Ich sehe auf Person Y auch die Rolle "Leitung Gruppe A", obwohl die Rolle "Leitung Gruppe A" nicht zur Inklusion im Abo beisteuert.

File: app/domain/export/tabular/people/group_roles/mailing_list.rb (hunk 3 above)
Comment (carlobeltrame) [reply app/domain/export/tabular/people/group_roles/mailing_list.rb:23]: Ausserdem ist das sowieso falsch: `s.subscriber_id` ist nicht zwingend die Gruppe in der die `role_types` vorkommen. Z.B. kann `subscriber_id` die ID des Dachverbands sein, aber die `role_types` können Rollen-Typen in Untergruppen des Dachverbands sein.
Bei den MailingLists werden die eigentlichen GruppenID/RollenTyp-Paare glaube ich nie wirklich berechnet, aber für die Anzeige werden sie mit diesem Code gruppiert: https://github.com/hitobito/hitobito/blob/master/app/models/subscription.rb#L78..L90
Dieser Code kann aber leider nicht 1:1 hier wiederverwendet werden, weil `Role::TypeList` nicht Gruppen-IDs sondern Gruppennamen ausspuckt. Daher müssten wir diese Logik hier nochmals nachprogrammieren.
